src/program.cpp

```diff
@@ -105,829 +63,148 @@ program::program(const program& p) { assign(p); }
 {
     os << "migraphx::program p;" << std::endl;
-    // cppcheck-suppress variableScope
-    unsigned long seed = 0;
-    print_program(*this, [&](auto ins, const auto& names) {
-        auto op = cpp_op_var(names.at(ins), ins);
-        if(ins->name().front() != '@')
-        {
-            os << "migraphx::op::" << ins->name() << " " << op << ";" << std::endl;
-            print_op_attributes(os, op, ins->get_operator());
-        }
-        os << "auto " << cpp_var_name(names.at(ins)) << " = ";
-        if(ins->name() == "@literal")
-        {
-            os << "p.add_literal(";
-            bool use_abs = false;
-            ins->get_literal().visit([&](auto v) {
-                use_abs = std::none_of(v.begin(), v.end(), [](auto x) { return x < 0; });
-            });
-            if(use_abs)
-                os << "migraphx::abs(";
-            os << "migraphx::generate_literal(";
-            print_cpp_shape(os, ins->get_shape());
-            os << ", " << seed << ")";
-            if(use_abs)
-                os << ")";
-            os << ");" << std::endl;
-            seed++;
-        }
-        else if(ins->name() == "@param")
-        {
-            std::string name = any_cast<builtin::param>(ins->get_operator()).parameter;
-            os << "p.add_parameter(" << enclose_name(name) << ",";
-            print_cpp_shape(os, ins->get_shape());
-            os << ");" << std::endl;
-        }
-        else
-        {
-            os << "p.add_instruction(" << op;
-            for(auto input : ins->inputs())
-            {
-                os << ", " << cpp_var_name(names.at(input));
-            }
-            os << ");" << std::endl;
-        }
-    });
+    main_module.print_cpp(os);
 }
 
 void program::dry_run(std::unordered_map<std::string, argument> params) const
 {
-    auto& ctx = this->impl->ctx;
-    generic_eval(*this, ctx, std::move(params), [](auto&&...) { return argument{}; });
+    main_module.dry_run(std::move(params));
 }
 
 void program::annotate(std::ostream& os, std::function<void(instruction_ref)> a) const
 {
-    print_program(*this, [&](auto ins, const auto& names) {
-        print_instruction(os, ins, names);
-        a(ins);
-        os << std::endl;
-    });
+    main_module.annotate(os, std::move(a));
 }
 
 program& program::sort()
 {
-    fix([&](auto self, auto ins) {
-        this->move_instruction(ins, this->begin());
-        for(auto child : ins->inputs())
-            self(child);
-    })(std::prev(this->end()));
-    assert(this->validate() == this->end());
+    main_module.sort();
     return *this;
 }
 
 bool operator==(const program& x, const program& y) { return to_string(x) == to_string(y); }
 
 std::ostream& operator<<(std::ostream& os, const program& p)
 {
-    print_program(p, [&](auto ins, const auto& names) {
-        print_instruction(os, ins, names);
-        os << std::endl;
-    });
+    const auto* mm = p.get_main_module();
+    os << *mm;
```

Comment:
This should print `main:` first, to show the instructions are coming from the main module.